salt/_modules/metalk8s_volumes.py

```diff
@@ -0,0 +1,430 @@
+
+    Returns:
+        None
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' metalk8s_volumes.format example-volume
+    """
+    _get_volume(name).format()
+
+
+# Volume {{{
+
+
+class Volume(object):
+    """Volume interface."""
+
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, volume):
+        self._volume = volume
+
+    @abc.abstractproperty
+    def exists(self):
+        """Does the backing storage device exists?"""
+        return
+
+    @abc.abstractmethod
+    def create(self):
+        """Create the backing storage device."""
+        return
+
+    @abc.abstractproperty
+    def is_finalized(self):
+        """Check if the backing storage device setup is finalized."""
+        return
+
+    @abc.abstractmethod
+    def finalize(self):
+        """Finalize the setup of the backing storage device."""
+        return
+
+    @abc.abstractproperty
+    def path(self):
+        """Path to the backing device."""
+        return
+
+    @property
+    def is_formatted(self):
+        """Check if the volume is already formatted."""
+        fs_type = self.get('spec.storageClassName.parameters.fsType')
+        return _get_fstype(self.path) == fs_type
+
+    def format(self, force=False):
+        """Format the volume.
+
+        The volume is formatted according to its StorageClass.
+        """
+        # Check that the backing device is not already formatted.
+        # Bail out if it is: we don't want data loss because of a typo…
+        if _get_fstype(self.path):
+            raise Exception('backing device `{}` already formatted'.format(
+                self.path
+            ))
+        params = self.get('spec.storageClassName.parameters')
+        # mkfs options, if any, are stored as JSON-encoded list.
+        options = json.loads(params.get('mkfsOptions', '[]'))
+        fs_type = params['fsType']
+        command = _mkfs(
+            self.path, fs_type, self.get('metadata.uid'), force, options
+        )
+        _run_cmd(' '.join(command))
+
+    def get(self, path):
+        """Return the Volume attribute `path` from the Volume dict."""
+        return functools.reduce(operator.getitem, path.split('.'), self._volume)
+
+
+# }}}
+# SparseLoopDevice {{{
+
+
+class SparseLoopDevice(Volume):
+    @property
+    def path(self):
+        return '/var/lib/metalk8s/storage/sparse/{}'.format(
+            self.get('metadata.name')
+        )
+
+    @property
+    def size(self):
+        return _quantity_to_bytes(self.get('spec.sparseLoopDevice.size'))
+
+    @property
+    def exists(self):
+        return (
+            os.path.isfile(self.path) and
+            os.path.getsize(self.path) == self.size
+        )
+
+    def create(self):
+        # Try to create a sparse file, don't clobber existing one!
+        open_flags = os.O_WRONLY | os.O_CREAT | os.O_EXCL
+        try:
+            with _open_fd(self.path, open_flags) as fd:
+                os.ftruncate(fd, self.size)
```

Comment:
The calculation of `self.size` should be moved *before* the file is opened/created.

Rationale: if `self.size` calculation fails (e.g., invalid format of the value), the `open` will still have happened, so the file will (zero-length) exist on disk. Now when a user would fix the size field, creation will bail out again (size mismatch).

Reply:
Good catch!! I'll move it before the `open`.

Reply:
Hum, we will still have the issue if the `truncate` fails no?

Instead of moving the calcultation, maybe I should have a try/finally around the `truncate` to `os.unlink` in case of error.

Reply:
But only if the error is not `EEXIST` or whatever, then... Otherwise it kinda defeats the purpose :D

Reply:
True, but what I had in mind was:

```python
        open_flags = os.O_WRONLY | os.O_CREAT | os.O_EXCL
        try:
            with _open_fd(self.sparse_file, open_flags) as fd:
                try:
                    os.ftruncate(fd, self.size)
                except Exception:
                    os.unlink(self.sparse_file)
                    raise
        except OSError as exn:
            raise Exception('cannot create sparse file at {}: {}'.format(
                self.path, exn
            ))
```
So it should be safe (the unlink only happen if `truncate` (or `self.size`), not for `open`).